### PR TITLE
Python script for database initialization with roles

### PR DESCRIPTION
**Pull request by geomatician**
This Python script creates a new Database with roles for SWOOP API / Caboose / Conductor, and an owner role. 
The script is added to the Swoop-DB docker image at `quay.io/element84/swoop-db` via the Dockerfile. This PR is linked with [this other PR](https://github.com/Element84/filmdrop-k8s-tf-modules/pull/33) that creates a KubernetesJob in Terraform that runs this script.

### CODE REVIEW

File: bin/db-initialization.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import os
+
+from buildpg import V, asyncpg, render
+
+DB_NAME = "test"
+
+
+def db_initialization():
+    async def setup():
+        dsn = "postgresql://postgres:password@postgres:5432/postgres"
```
Comment (hectormachin) [inline bin/db-initialization.py]: Could we pass these credentials via ENV variables?

File: Dockerfile [hunk 2]
```diff
@@ -33,3 +33,5 @@ ENV PGUSER: "${PGUSER:-postgres}"
 # copy the python venv into this output image and add it's bin to the path
 COPY --from=APP /opt/swoop/db/swoop-db-venv /opt/swoop/db/swoop-db-venv
 ENV PATH=/opt/swoop/db/swoop-db-venv/bin:$PATH
+
+COPY bin/db-initialization.py home/db-initialization.py
```
Comment (hectormachin) [inline Dockerfile]: Can we change the location of this script outside the home directory? Maybe somewhere within /opt/swoop/db?

**Review by hectormachin — CHANGES_REQUESTED** [review]

**Review by hectormachin — APPROVED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -23,7 +23,7 @@ RUN set -x && \
     trap "rm -rf '$tmp'" EXIT && \
     cd "$tmp" && \
     curl -fsSL https://github.com/theory/pgtap/archive/refs/tags/v${PGTAP_VERSION}.tar.gz \
-        -o pgtap.tar.gz && \
+    -o pgtap.tar.gz && \
```
Comment (jkeifer) [inline Dockerfile]: This is a continuation of the previous line, not another command, so I don't think it should be unindented.

File: bin/db-initialization.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+DB_NAME = os.environ["DATABASE_TO_CREATE"]
+
+
+def db_initialization():
+    async def setup():
+        user = os.environ["POSTGRES_DEFAULT_USERNAME"]
+        password = os.environ["POSTGRES_DEFAULT_PASSWORD"]
+        dsn = f"postgresql://{user}:{password}@postgres:5432/postgres"
+        conn = await asyncpg.connect(dsn=dsn)
```
Comment (jkeifer) [inline bin/db-initialization.py]: See https://www.postgresql.org/docs/current/libpq-envars.html. We should be using these environment variables alone to define the connection parameters. By leaving the connection parameters to environment variables in this way we automatically support all connection parameters supported by libpq _without code changes_.

These lines can then simply become:

```suggestion
        conn = await asyncpg.connect(database="")
```

Note that the definition of `database` here overrides anything set in the `PGDATABASE` env var, so we can use that var for the name of the database we want to create, which means we don't have to specify a "special" var (`DATABASE_TO_CREATE`) for this script.

File: bin/db-initialization.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        # Create Owner Role
+
+        owner_role_un = os.environ["OWNER_ROLE_USER"]
+
+        q, p = render(
+            """
+                CREATE ROLE :un WITH CREATEDB CREATEROLE LOGIN PASSWORD ':pw';
```
Comment (jkeifer) [inline bin/db-initialization.py]: This user should not have `CREATEDB` or `CREATEROLE` permissions.

File: bin/db-initialization.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                CREATE ROLE :un WITH LOGIN IN ROLE :ir PASSWORD ':pw';
+            """,
+            un=V(os.environ["CONDUCTOR_ROLE_USER"]),
+            pw=V(os.environ["CONDUCTOR_ROLE_PASS"]),
+            ir=V(owner_role_un),
+        )
+
+        await conn.execute(q, *p)
+
+        # Create database with owner role as owner
```
Comment (jkeifer) [inline bin/db-initialization.py]: All of this is the same code, it can be collapsed into a loop.

File: bin/db-initialization.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            pw=V(os.environ["OWNER_ROLE_PASS"]),
+        )
+
+        await conn.execute(q, *p)
+
+        # Create Swoop API Role
+
+        q, p = render(
+            """
+                CREATE ROLE :un WITH LOGIN IN ROLE :ir PASSWORD ':pw';
```
Comment (jkeifer) [inline bin/db-initialization.py]: Idempotency is important; how do you ensure this script can run again without errors?

To expand upon this, if something goes wrong while running this script a first time, we want to be able to run it again and have it pick up where it left off, not error out when it was only partially applied.

File: bin/db-initialization.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
+import asyncio
+import os
+
+from buildpg import V, asyncpg, render
+
+DB_NAME = os.environ["DATABASE_TO_CREATE"]
```
Comment (jkeifer) [inline bin/db-initialization.py]: Generally speaking, you should avoid reading env vars as part of module initialization. It makes it difficult if you want to override them after importing.

File: bin/db-initialization.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
+import asyncio
+import os
+
+from buildpg import V, asyncpg, render
```
Comment (jkeifer) [inline bin/db-initialization.py]: You aren't using the special `buildpg` extensions to `asyncpg`, so it seems like this should be a vanilla `asyncpg` import.

File: bin/db-initialization.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+DB_NAME = os.environ["DATABASE_TO_CREATE"]
+
+
+def db_initialization():
```
Comment (jkeifer) [inline bin/db-initialization.py]: Seems like this should be the async function, and just wrap the call in the main check instead of in this function.

Comment (jkeifer) [reply bin/db-initialization.py]: Also, this function is really long. It doesn't do a whole lot, but it doesn't look like it. I think breaking it down in to smaller, self-documenting functions would make the code more obvious.

File: bin/db-initialization.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
+import asyncio
```
Comment (jkeifer) [inline bin/db-initialization.py:18]: It's a good idea to make scripts executable and add a shebang so they can be run directly.

It also seems like we need some documentation in this script explaining how to run it.

**Review by jkeifer — CHANGES_REQUESTED** [review]
This is a good starting point, but I think it could use some refactoring and a few functional changes.

As an example of where I would take this, see https://github.com/Element84/swoop-db/pull/24/. Not that you have to use that example (but you can if you want).

File: Dockerfile (hunk 3 above)
Comment (hectormachin) [reply Dockerfile]: resolved

**Review by hectormachin — APPROVED** [review]

**Pull request merged**